File: backend/app/models/message.py
"""Message models for Redis communication."""

from typing import Any, Dict, Literal, Optional
from pydantic import BaseModel


class BaseMessage(BaseModel):
    """Base message structure."""
    id: str
    type: str
    timestamp: int
    source: Literal["mcp", "backend", "frontend"]
    target: str
    component: Optional[Literal["todo", "backlog"]] = None
    payload: Dict[str, Any]


class TodoActionMessage(BaseMessage):
    """Todo action message."""
    type: Literal["todo_action"] = "todo_action"
    target: Literal["todo_component"] = "todo_component"


class SSEEvent(BaseModel):
    """SSE event structure."""
    event: str
    data: Dict[str, Any]

File: shared/types/message.ts
/**
 * Shared message types for communication between components
 */

export interface BaseMessage {
  id: string;
  type: string;
  timestamp: number;
  source: 'mcp' | 'backend' | 'frontend';
  target: string;
  component?: 'todo' | 'backlog' | 'terminal' | 'approval';
  payload: any;
}

export interface TodoActionMessage extends BaseMessage {
  type: 'todo_action';
  target: 'todo_component';
  payload: {
    action: 'add' | 'delete' | 'update' | 'toggle';
    todoId?: string;
    data?: {
      title?: string;
      description?: string;
      completed?: boolean;
    };
  };
}

export interface ComponentStateMessage extends BaseMessage {
  type: 'component_state';
  payload: {
    componentId: string;
    state: any;
  };
}

export interface ApprovalRequestMessage extends BaseMessage {
  type: 'approval_request';
  target: 'approval_component';
  payload: {
    session_id: string;
    function_call_id: string;
    description: string;
  };
}

export type MessageType = 'todo_action' | 'backlog_action' | 'terminal_action' | 'approval_request';

export interface SSEEvent {
  event: string;
  data: {
    messageId?: string;
    type: string;
    payload: any;
  };
}
